third_party/nvfuser/csrc/scheduler/normalization_utils.cpp

```diff
@@ -494,5 +502,196 @@ std::optional<GridOuterNormalizationParams> getGridOuterNormalizationParams(
+      //   T5[b, b, r] = reduction(T4)
+      //
+      // After projection, it becomes
+      //
+      //   T0[I]
+      //   T1[b b I] = broadcast(T0)
+      //   T2[b b r] = reduction(T1)
+      //   T3[b b b] = broadcast(T2)
+      //   T6[b b I] = broadcast(T0)
+      //   T4[b, b, I] = T6 + T3
+      //   T5[b, b, r] = reduction(T4)
+      //
+      // During schedule, we need to propagate from T2 to T5. However, in the
+      // resulting DAG, neither the propagation path T2->T3->T4->T5 nor
+      // T2->T1->T0->T6->T4->T5 works because they both have missing root
+      // domain. But adding `T7 = T1 + T6` creates a new propagation path
+      // `T2->T1->T7->T6->T4->T5` which has all root domain information.
+      // See FusionBroadcastPersistentReduction_CUDA for an example
+      dummy_outputs.emplace_back(add(buffer_replicate, buffer));
+      ir_utils::replaceValInExpr(use->definition(), buffer, buffer_replicate);
+    }
+  }
+  return dummy_outputs;
+}
+
+std::unordered_set<TensorView*> fixUpInvalidPersistentBuffers(Fusion* fusion) {
+  auto persistent_info = scheduler_utils::persistentBuffers(fusion);
+  const auto& persistent_buffers = persistent_info.persistent_buffers;
+  const auto& persistent_resolution_points =
+      persistent_info.persistent_buffer_resolution_points;
+
+  std::unique_ptr<ConcretizedBroadcastDomains> concretize_info;
+
+  std::unordered_set<TensorView*> recomputed_tvs;
+
+  bool recompute_done = false;
+
+  for (auto buffer_i : c10::irange(persistent_buffers.size())) {
+    auto buffer = persistent_buffers.at(buffer_i);
+
+    // Check if this buffer needs to be recomputed
+    bool need_recomputation = false;
+
+    for (const auto i : c10::irange(
+             buffer->getComputeAtPosition(),
+             buffer->domain()->domain().size())) {
+      auto axis = buffer->axis(i);
+
+      // Unresolved data dependency could exist if:
+      // - Parallelized by tidx and stored on Local
+      // - Parallelized by bidx and stored on Local or Shared
+      if ((axis->isThreadDim() &&
+           buffer->getMemoryType() == MemoryType::Local) ||
+          (axis->isBlockDim() &&
+           (buffer->getMemoryType() == MemoryType::Local ||
+            buffer->getMemoryType() == MemoryType::Shared))) {
+        if (!concretize_info) {
+          concretize_info =
+              std::make_unique<ConcretizedBroadcastDomains>(fusion);
+        }
+
+        // concretize_info.isConcretized(axis) means the axis is a
+        // concreteized broadcast domain
+        if (!concretize_info->isConcretized(axis)) {
+          continue;
+        }
+
+        need_recomputation = true;
+        break;
+      }
+    }
+
+    if (!need_recomputation) {
+      continue;
+    }
+
+    const auto& resolution_points = persistent_resolution_points.at(buffer_i);
+
+    const auto persistent_use_of_buffer =
+        getPersistentUseOfBuffer(buffer, resolution_points);
+
+    for (auto use : persistent_use_of_buffer) {
```

Comment:
So this is just adding a new expression for the first use of the persistent buffer?

Reply:
This just follows the same logic as the buffer projection. It replicates all the dependent expressions from inputs to the first use of the persistent buffer.

Reply:
That's what I thought, I'm just forgetting how as it looks like it's only recomputing one use.